File: src/urlNormalize.ts
/** 与后端 ``normalize_original_url`` 对齐，用于收藏去重与前端比对 */

export function normalizeOriginalUrl(url: string): string {
  let raw = (url || '').trim()
  if (!raw) return ''
  if (raw.startsWith('//')) raw = `https:${raw}`
  try {
    const u = new URL(raw)
    const scheme = (u.protocol.replace(':', '') || 'https').toLowerCase()
    const host = u.host.toLowerCase()
    let path = u.pathname.replace(/\/+$/, '')
    if (!path) path = '/'
    return `${scheme}://${host}${path}${u.search}`
  } catch {
    return ''
  }
}

File: backend/url_normalize.py
"""收藏去重：将商品详情 URL 规范为稳定字符串（与搜索合并去重思路一致）。"""

from __future__ import annotations

from urllib.parse import urlparse, urlunparse


def normalize_original_url(url: str) -> str:
    """
    scheme/host 小写，path 去尾斜杠，保留 query（部分站点用查询区分商品）。
    """
    raw = (url or "").strip()
    if not raw:
        return ""
    parsed = urlparse(raw)
    scheme = (parsed.scheme or "https").lower()
    netloc = parsed.netloc.lower()
    path = (parsed.path or "").rstrip("/")
    if not path:
        path = "/"
    return urlunparse((scheme, netloc, path, "", parsed.query, ""))
